From 5857a851748407852b13c9d6d5e0ffd94e504d60 Mon Sep 17 00:00:00 2001
From: rmp22 <195054967+rmp22@users.noreply.github.com>
Date: Wed, 27 Aug 2025 18:26:05 +0800
Subject: [PATCH] [bugfix] add perf activity anim override

fix for mtk and low-end devices perf regression caused by fixed dp translate and fade animation.
This version applies the override unconditionally without checking system properties.

---
 core/java/android/view/animation/AnimationUtils.java | 112 ++++++++++++++++++
 1 file changed, 112 insertions(+)

diff --git a/core/java/android/view/animation/AnimationUtils.java b/core/java/android/view/animation/AnimationUtils.java
index 8a407c3252f0c..9189f8e65885b 100644
--- a/core/java/android/view/animation/AnimationUtils.java
+++ b/core/java/android/view/animation/AnimationUtils.java
@@ -33,10 +33,13 @@
 import android.ravenwood.annotation.RavenwoodIgnore;
 import android.ravenwood.annotation.RavenwoodKeepPartialClass;
 import android.util.AttributeSet;
+import android.util.DisplayMetrics;
 import android.util.TimeUtils;
 import android.util.Xml;
 import android.view.InflateException;
 
+import com.android.internal.R;
+
 import org.xmlpull.v1.XmlPullParser;
 import org.xmlpull.v1.XmlPullParserException;
 
@@ -223,7 +226,17 @@
      */
     public static Animation loadAnimation(Context context, @AnimRes int id)
             throws NotFoundException {
-
+        ActivityAnimations.maybeInit(context);
+        switch (id) {
+            case R.anim.activity_open_enter:
+                return ActivityAnimations.getOpenEnter();
+            case R.anim.activity_open_exit:
+                return ActivityAnimations.getOpenExit();
+            case R.anim.activity_close_enter:
+                return ActivityAnimations.getCloseEnter();
+            case R.anim.activity_close_exit:
+                return ActivityAnimations.getCloseExit();
+        }
         XmlResourceParser parser = null;
         try {
             parser = context.getResources().getAnimation(id);
@@ -504,4 +517,99 @@
         }
         return interpolator;
     }
+    /** @hide */
+    public final class ActivityAnimations {
+
+        private static Animation sOpenEnter;
+        private static Animation sOpenExit;
+        private static Animation sCloseEnter;
+        private static Animation sCloseExit;
+
+        private static Interpolator sFastOutExtraSlowInInterpolator;
+
+        private static final float DISTANCE = 0.1f;
+
+        private ActivityAnimations() {}
+
+        /** @hide */
+        public static void maybeInit(Context context) {
+            if (sFastOutExtraSlowInInterpolator == null) {
+                sFastOutExtraSlowInInterpolator = AnimationUtils.loadInterpolator(
+                        context, R.interpolator.fast_out_extra_slow_in);
+            }
+        }
+
+        private static class ActivityAnimFactory {
+            private float fromX = 0f, toX = 0f;
+            private long duration = 350L;
+
+            public ActivityAnimFactory fromX(float ratio) {
+                this.fromX = ratio;
+                return this;
+            }
+
+            public ActivityAnimFactory toX(float ratio) {
+                this.toX = ratio;
+                return this;
+            }
+
+            public Animation build() {
+                AnimationSet animationSet = new AnimationSet(false);
+                TranslateAnimation slide = new TranslateAnimation(
+                        Animation.RELATIVE_TO_SELF, fromX,
+                        Animation.RELATIVE_TO_SELF, toX,
+                        Animation.RELATIVE_TO_SELF, 0f,
+                        Animation.RELATIVE_TO_SELF, 0f
+                );
+                slide.setDuration(duration);
+                slide.setInterpolator(sFastOutExtraSlowInInterpolator);
+                animationSet.addAnimation(slide);
+                return animationSet;
+            }
+        }
+
+        /** @hide */
+        public static Animation getOpenEnter() {
+            if (sOpenEnter == null) {
+                sOpenEnter = new ActivityAnimFactory()
+                        .fromX(1.0f)
+                        .toX(0.0f)
+                        .build();
+            }
+            return sOpenEnter;
+        }
+
+        /** @hide */
+        public static Animation getOpenExit() {
+            if (sOpenExit == null) {
+                sOpenExit = new ActivityAnimFactory()
+                        .fromX(0.0f)
+                        .toX(-DISTANCE)
+                        .build();
+            }
+            return sOpenExit;
+        }
+
+        /** @hide */
+        public static Animation getCloseEnter() {
+            if (sCloseEnter == null) {
+                sCloseEnter = new ActivityAnimFactory()
+                        .fromX(-DISTANCE)
+                        .toX(0.0f)
+                        .build();
+            }
+            return sCloseEnter;
+        }
+
+        /** @hide */
+        public static Animation getCloseExit() {
+            if (sCloseExit == null) {
+                sCloseExit = new ActivityAnimFactory()
+                        .fromX(0.0f)
+                        .toX(1.0f)
+                        .build();
+            }
+            return sCloseExit;
+        }
+    }
 }
 